File: src/integration/spring/src/jvmTest/java/community/flock/wirespec/integration/spring/java/it/client/WebClientIntegrationTest.java
package community.flock.wirespec.integration.spring.java.it.client;

import community.flock.wirespec.integration.spring.java.application.Application;
import community.flock.wirespec.integration.spring.java.generated.endpoint.*;
import community.flock.wirespec.integration.spring.java.generated.model.ApiResponse;
import community.flock.wirespec.integration.spring.java.generated.model.Pet;
import community.flock.wirespec.integration.spring.java.generated.model.UploadFileRequestBody;
import community.flock.wirespec.integration.spring.java.generated.model.UploadFileRequestBodyJson;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;

import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT, classes = {Application.class, WirespecPetstoreWebClient.class})
@EnableConfigurationProperties
public class WebClientIntegrationTest {

    @Autowired
    private WirespecPetstoreWebClient wirespecPetstoreWebClient;

    @Test
    public void crudWebClientIntegrationTest() throws ExecutionException, InterruptedException {
        Pet pet = new Pet(
                Optional.of(1L),
                "Dog",
                Optional.empty(),
                List.of(),
                Optional.empty(),
                Optional.empty()
        );

        AddPet.Response<?> addPetResponse = wirespecPetstoreWebClient.addPet(new AddPet.Request(pet)).get();
        assertEquals(new AddPet.Response200(Optional.of(100), pet), addPetResponse);

        Pet updatedPet = new Pet(
                pet.id(),
                "Cat",
                pet.category(),
                pet.photoUrls(),
                pet.tags(),
                pet.status()
        );
        UpdatePet.Response<?> updatePetResponse = wirespecPetstoreWebClient.updatePet(new UpdatePet.Request(updatedPet)).get();
        assertEquals(new UpdatePet.Response200(updatedPet), updatePetResponse);

        GetPetById.Response<?> getPetResponse = wirespecPetstoreWebClient.getPetById(new GetPetById.Request(pet.id().orElseThrow())).get();
        assertEquals(new GetPetById.Response200(updatedPet), getPetResponse);

        DeletePet.Response<?> deletePetResponse = wirespecPetstoreWebClient.deletePet(new DeletePet.Request(pet.id().get(), Optional.empty())).get();
        assertEquals(new DeletePet.Response400(), deletePetResponse);
    }

    @Test
    public void queryParameters() throws ExecutionException, InterruptedException {
        FindPetsByTags.Response<?> queryParamResponse = wirespecPetstoreWebClient.findPetsByTags(
                new FindPetsByTags.Request(Optional.of(List.of("Smilodon", "Dodo", "Mammoth")))
        ).get();
        assertEquals(new FindPetsByTags.Response200(List.of()), queryParamResponse);
    }

    @Test
    public void multipartFormData() throws ExecutionException, InterruptedException {
        UploadFile.Request request = new UploadFile.Request(
                1L,
                Optional.of("metadata"),
                new UploadFileRequestBody(
                        Optional.of("metadata"),
                        Optional.of("data".getBytes()),
                        Optional.of(new UploadFileRequestBodyJson(Optional.of("bar")))
                )
        );
        UploadFile.Response<?> response = wirespecPetstoreWebClient.uploadFile(request).get();
        assertEquals(
                new UploadFile.Response200(new ApiResponse(Optional.of(200), Optional.of("type"), Optional.of("metadata"))),
                response
        );
    }
}

File: src/integration/spring/src/jvmTest/kotlin/community/flock/wirespec/integration/spring/kotlin/it/client/WebClientIntegrationTest.kt
package community.flock.wirespec.integration.spring.kotlin.it.client

import community.flock.wirespec.integration.spring.kotlin.application.Application
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.AddPet
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.DeletePet
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.FindPetsByTags
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.GetPetById
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.UpdatePet
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.UploadFile
import community.flock.wirespec.integration.spring.kotlin.generated.model.ApiResponse
import community.flock.wirespec.integration.spring.kotlin.generated.model.Pet
import community.flock.wirespec.integration.spring.kotlin.generated.model.UploadFileRequestBody
import community.flock.wirespec.integration.spring.kotlin.generated.model.UploadFileRequestBodyJson
import community.flock.wirespec.openapi.json
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.boot.test.context.SpringBootTest
import kotlin.test.Test

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT, classes = [Application::class, WirespecPetstoreWebClient::class])
@EnableConfigurationProperties
class WebClientIntegrationTest {
    @Autowired
    lateinit var wirespecPetstoreWebClient: WirespecPetstoreWebClient

    @Test
    fun `CRUD WebClient integration test`(): Unit = runBlocking {
        val pet =
            Pet(
                id = 1,
                name = "Dog",
                photoUrls = listOf(),
                category = null,
                tags = null,
                status = null,
            )

        val addPetResponse = wirespecPetstoreWebClient.addPet(AddPet.Request(pet))
        assertEquals(AddPet.Response200(pet), addPetResponse)

        val updatedPet = pet.copy(name = "Cat")
        val updatePetResponse = wirespecPetstoreWebClient.updatePet(UpdatePet.Request(updatedPet))
        assertEquals(UpdatePet.Response200(updatedPet), updatePetResponse)

        val getPetResponse = wirespecPetstoreWebClient.getPetById(GetPetById.Request(pet.id!!))
        assertEquals(GetPetById.Response200(updatedPet), getPetResponse)

        val deletePetResponse = wirespecPetstoreWebClient.deletePet(DeletePet.Request(pet.id, null))
        assertEquals(DeletePet.Response400(Unit), deletePetResponse)
    }

    @Test
    fun `query parameters`() = runBlocking {
        val queryParamResponse =
            wirespecPetstoreWebClient.findPetsByTags(
                FindPetsByTags.Request(listOf("Smilodon", "Dodo", "Mammoth")),
            )
        assertEquals(FindPetsByTags.Response200(emptyList()), queryParamResponse)
    }

    @Test
    fun `multipart form data`() = runBlocking {
        val request = UploadFile.Request(
            petId = 1,
            additionalMetadata = "metadata",
            body = UploadFileRequestBody(
                additionalMetadata = "metadata",
                file = "data".toByteArray(),
                json = UploadFileRequestBodyJson(foo = "bar"),
            ),
        )
        val response = wirespecPetstoreWebClient.uploadFile(request)
        assertEquals(
            UploadFile.Response200(ApiResponse(code = 200, type = "type", message = "metadata")),
            response,
        )
    }
}
